lib/Conversion/TorchOnnxToTorch/OnnxGruExpander.cpp

```diff
@@ -0,0 +1,431 @@
+
+    Type hTyUnsqueezed = b.getType<ValueTensorType>(
+        llvm::SmallVector<int64_t>{1, batch_size, hidden_size}, hTy.getDtype());
+    Value H_new_unsqueezed =
+        b.create<AtenUnsqueezeOp>(hTyUnsqueezed, H_new, cstZero);
+
+    auto loopIndexPlusOne = b.create<AtenAddIntOp>(intType, loopIndex, cstOne);
+    Value Y_new =
+        b.create<AtenSliceScatterOp>(yTy, Y_prev, H_new_unsqueezed, cstZero,
+                                     loopIndex, loopIndexPlusOne, cstOne);
+
+    b.create<PrimLoopConditionOp>(loopConditionTrue,
+                                  ValueRange({Y_new, H_new}));
+  }
+
+  GruLayerOutput output;
+  output.Y = loop.getResult(0);
+  output.Y_h = loop.getResult(1);
+
+  return output;
+}
+
+LogicalResult OnnxGruExpander(OpBinder binder,
+                              ConversionPatternRewriter &rewriter) {
+  Location loc = binder.getLoc();
+  mlir::ImplicitLocOpBuilder b(loc, rewriter);
+
+  // Binding arguments
+  ValueTensorType yTy, Y_hType;
+  if (binder.tensorResultTypeAtIndex(yTy, 0) ||
+      binder.tensorResultTypeAtIndex(Y_hType, 1)) {
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "At least one output must be present");
+  }
+
+  Value X, W, R, B, initial_h, sequence_lens;
+  if (binder.tensorOperandAtIndex(X, 0) || binder.tensorOperandAtIndex(W, 1) ||
+      binder.tensorOperandAtIndex(R, 2))
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Missing required input tensor");
+
+  if (binder.tensorOperandAtIndex(B, 3))
+    B = b.create<AtenZerosOp>(W.getType(), W);
+
+  int64_t hidden_size;
+  if (binder.s64IntegerAttr(hidden_size, "hidden_size"))
+    return rewriter.notifyMatchFailure(
+        binder.op, "Missing required attribute hidden_size");
+
+  auto xTy = cast<ValueTensorType>(X.getType());
+  auto wTy = cast<ValueTensorType>(W.getType());
+
+  // Setting up activations
+  GruActivations activations;
+  activations.f = "Sigmoid";
+  activations.g = "Tanh";
+
+  llvm::SmallVector<std::string> activationsList;
+  if (!binder.stringArrayAttr(activationsList, "activations") &&
+      activationsList.size() == 2) {
+    activations.f = activationsList[0];
+    activations.g = activationsList[1];
+  } else if (activationsList.size() > 0) {
+    return rewriter.notifyMatchFailure(
+        binder.op, "Unsupported number of activation functions");
+  }
+
+  // Other attributes
+  int64_t layout;
+  if (binder.s64IntegerAttr(layout, "layout", 0))
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Unsupported layout attribute type.");
+
+  std::string direction;
+  if (!binder.customOpNameStringAttr(direction, "direction", "forward") &&
+      direction != "forward")
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Unsupported direction attribute value");
+
+  int64_t num_directions = 1 + (direction == "bidirectional");
```

Comment:
I think Rob had a comment about lines like this. Maybe change to a ternary expression.